src/module-loader/dotnet/src/MorganStanley.ComposeUI.ModuleLoader/AggregateModuleCatalog.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Morgan Stanley makes this available to you under the Apache License,
+// Version 2.0 (the "License"). You may obtain a copy of the License at
+// 
+//      http://www.apache.org/licenses/LICENSE-2.0.
+// 
+// See the NOTICE file distributed with this work for additional information
+// regarding copyright ownership. Unless required by applicable law or agreed
+// to in writing, software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions
+// and limitations under the License.
+
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace MorganStanley.ComposeUI.ModuleLoader;
+
+internal class AggregateModuleCatalog : IModuleCatalog
+{
+    private readonly IEnumerable<IModuleCatalog> _moduleCatalogs;
+    private readonly ILogger _logger;
+
+    public AggregateModuleCatalog(
+        IEnumerable<IModuleCatalog> moduleCatalogs,
+        ILogger? logger = null)
+    {
+        _moduleCatalogs = moduleCatalogs;
+        _logger = logger ?? NullLogger.Instance;
+        _lazyEnumerateModules = new Lazy<Task>(EnumerateModulesCore);
+    }
+
+    public async Task<IModuleManifest> GetManifest(string moduleId)
+    {
+        await EnumerateModules();
+
+        return _moduleIdToCatalog.TryGetValue(moduleId, out var catalog)
+            ? await catalog.GetManifest(moduleId)
+            : throw new ModuleNotFoundException(moduleId);
+    }
+
+    public async Task<IEnumerable<string>> GetModuleIds()
+    {
+        await EnumerateModules();
+
+        return _moduleIdToCatalog.Keys;
+    }
+
+    private readonly Lazy<Task> _lazyEnumerateModules;
+    private readonly ConcurrentDictionary<string, IModuleCatalog> _moduleIdToCatalog = new();
+
+    private Task EnumerateModules() => _lazyEnumerateModules.Value;
+
+    private async Task EnumerateModulesCore()
```

Comment:
[ZKRobi](https://github.com/ZKRobi) [4 days ago] --Old version is async fetching using LINQ.(https://github.com/morganstanley/ComposeUI/pull/569#discussion_r1555987809)
This piece of code seems correct, but it's quite hard to understand. I'd separate it to fetch the moduleIds in parallel (it might even be more efficient that way), concatenate the results in order of registration, and do the resolving logic on this.

Reply:
Resolved as it was resolved on the original PR.